Adding Codespaces Documentation (#1144)
python/devguide: getting-started/setup-building.rst

.. _issue tracker: https://github.com/python/cpython/issues

.. _using-codespaces:

> Contribute Using GitHub Codespaces

Reviewer: According to the devguide docs style guide, headings should be written using "sentence case":

https://devguide.python.org/documentation/style-guide/#capitalization

> In the Python documentation, the use of sentence case in section titles is preferable, but consistency within a unit is more important than following this rule. If you add a section to a chapter where most sections are in title case, you can either convert all titles to sentence case or use the dominant style in the new section title.

Reviewer 2: ```suggestion
Contribute using GitHub Codespaces
```

Author: Thanks for pointing that out! I'll go through and adjust the titles to match!

Author: Actually, thanks @pradyunsg for the suggestions to do just that 😄

Reviewer: Thanks! There's a lot of guidelines; it's hard to remember all of them. I misstep all the time 😄

==================================

.. _codespaces-whats-codespaces:

> What is Github Codespaces
> -------------------------

Reviewer 3: ```suggestion
What is GitHub Codespaces?
--------------------------
```

If you'd like to start contributing to CPython without needing to set up a local developer environment, you can use `Github Codespaces <https://github.com/features/codespaces>`_.
Codespaces is a cloud-based development environment offered by GitHub that allows developers to write, build, test, and debug code directly within their web browser or in Visual Studio Code (VS Code).

> To help you get started CPython contains a `devcontainer folder <https://github.com/python/cpython/tree/main/.devcontainer>`_ with a json file that provides consistent and versioned codespace configurations for all users of the project.

Reviewer 3: ```suggestion
To help you get started CPython contains a `devcontainer folder <https://github.com/python/cpython/tree/main/.devcontainer>`_ with a JSON file that provides consistent and versioned codespace configurations for all users of the project.
```

> To help you get started CPython contains a `devcontainer folder <https://github.com/python/cpython/tree/main/.devcontainer>`_ with a JSON file that provides consistent and versioned codespace configurations for all users of the project.

Reviewer 4: ```suggestion
To help you get started, CPython contains a `devcontainer folder <https://github.com/python/cpython/tree/main/.devcontainer>`_ with a JSON configuration file that provides consistent and versioned codespace configurations for all users of the project.
```

> It also contains a Dockerfile that allows you to set up the same environment but locally in a Docker container if you'd prefer.

Reviewer 4: ```suggestion
It also contains a Dockerfile that allows you to set up the same environment but locally in a Docker container if you'd prefer to use that directly.
```

.. _codespaces-create-a-codespace:

> Create A CPython Codespace

Reviewer 2: ```suggestion
Create a CPython codespace
```

--------------------------

Here are the basic steps needed to contribute a patch using Codespaces.
You first need to navigate to the `CPython repo <https://github.com/python/cpython>`_ hosted on GitHub.

Then you will need to:

> 1. Click the green :guilabel:`Code` button and choose the ``codespaces`` tab.
> 2. Press the green :guilabel:`Create codespace on main` button.

Reviewer 4: ```suggestion
1. Press the ``,`` key to launch the codespace setup screen for the current branch (alternatively, click the green :guilabel:`Code` button and choose the ``codespaces`` tab and then press the green :guilabel:`Create codespace on main` button).
```

3. A screen should appear that lets you know your codespace is being set up. (Note: Since the CPython devcontainer is provided, codespaces will use the configuration it specifies.)

> 4. VS Code will open inside your web browser, already linked up with your code and a terminal to the remote codespace.

Reviewer 4: ```suggestion
4. A `web version of VS Code <https://vscode.dev/>`_ will open inside your web browser, already linked up with your code and a terminal to the remote codespace where CPython and its documentation have already been built.
```

> 3. A screen should appear that lets you know your codespace is being set up. (Note: Since the CPython devcontainer is provided codespaces will use the configurations it specifies.)
> 4. VS Code will open inside of your web browser, already linked up with your code and a terminal to the remote codespace.
> 5. Use the terminal with the usual git commands to create a new branch, commit and push your changes once you're ready!

Reviewer 3: ```suggestion
3. A screen should appear that lets you know your codespace is being set up. (Note: Since the CPython devcontainer is provided, codespaces will use the configuration it specifies.)
4. VS Code will open inside your web browser, already linked up with your code and a terminal to the remote codespace.
5. Use the terminal with the usual Git commands to create a new branch, commit and push your changes once you're ready!
```

If you close your repository and come back later you can always resume your codespace by navigating to the CPython repo, selecting the codespaces tab and selecting your most recent codespaces session.
You should then be able to pick up from where you left off!

.. _codespaces-use-locally:

Use Codespaces Locally
-----------------------

On the bottom left side of the codespace screen you will see a green or grey square that says :guilabel:`Codespaces`.

> You can click on this for additional options. If you prefer working on a locally installed copy of VS Code you can select the option ``Open in VS Code``.

Reviewer 3: ```suggestion
You can click this for additional options. If you prefer working in a locally installed copy of VS Code you can select the option ``Open in VS Code``.
```

> You will still be working on the remote codespace instance, thus utilising the remote instance compute power.
> The compute power may be a much higher spec than your local machine which can be helpful.

Reviewer 3: ```suggestion
You will still be working on the remote codespace instance, thus using the remote instance's compute power.
The compute power may be a much higher spec than your local machine which can be helpful.
```

.. _issue tracker: https://github.com/python/cpython/issues

.. _using-codespaces:

Contribute using GitHub Codespaces
==================================

.. _codespaces-whats-codespaces:

What is GitHub Codespaces?
--------------------------

> set up the same environment but locally in a docker container if you'd prefer th

Reviewer: There's something off with the start of this sentence :)

Reviewer 3: It can be deleted, the full sentence is at line 682 below :)
